packages/astro/src/core/build/generate.ts

```diff
@@ -196,58 +198,35 @@ export async function generatePages(opts: StaticBuildOptions, internals: BuildIn
 		}
 	}
 
-	const staticImageList = getStaticImageList();
+	logger.info(null, dim(`Completed in ${getTimeStat(timer, performance.now())}.\n`));
 
-	if (staticImageList.size)
+	const staticImageList = getStaticImageList();
+	if (staticImageList.size) {
 		logger.info(null, `\n${bgGreen(black(` generating optimized images `))}`);
-	let count = 0;
-	for (const imageData of staticImageList.entries()) {
-		count++;
-		await generateImage(
-			pipeline,
-			imageData[1].options,
-			imageData[1].path,
-			count,
-			staticImageList.size
-		);
-	}
-
-	delete globalThis?.astroAsset?.addStaticImage;
 
-	await runHookBuildGenerated({
-		config: opts.settings.config,
-		logger: pipeline.getLogger(),
-	});
+		const totalCount = Array.from(staticImageList.values())
+			.map((x) => x.size)
+			.reduce((a, b) => a + b, 0);
+		const cpuCount = os.cpus().length;
+		const assetsCreationEnvironment = await prepareAssetsGenerationEnv(pipeline, totalCount);
+		const queue = new PQueue({ concurrency: cpuCount });
 
-	logger.info(null, dim(`Completed in ${getTimeStat(timer, performance.now())}.\n`));
-}
+		const assetsTimer = performance.now();
+		for (const [originalPath, transforms] of staticImageList) {
+			await generateImagesForPath(originalPath, transforms, assetsCreationEnvironment, queue);
+		}
```

Comment:
A question someone might have: Would make this part concurrent help?

No (at least as far as I can tell), it doesn't really matter if the queue is full of variants from one image vs variants of multiple images.